algorithms/kernel/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i

```diff
@@ -0,0 +1,933 @@
+            }
+
+            size_t nNodesNewLevel;
+            DAAL_CHECK_STATUS_VAR(_treeLevelBuildHelper.getNumOfSplitNodes(nodeList, nNodes, nNodesNewLevel));
+
+            if (nNodesNewLevel)
+            {
+                /*there are split nodes -> next level is required*/
+                nNodesNewLevel *= 2;
+
+                auto nodeListNewLevel = context.allocate(TypeIds::id<int>(), nNodesNewLevel * TreeLevel::_nNodeSplitProps, &status);
+                auto impListNewLevel =
+                    context.allocate(TypeIds::id<algorithmFPType>(), nNodesNewLevel * (TreeLevel::_nNodeImpProps + _nClasses), &status);
+                DAAL_CHECK_STATUS_VAR(status);
+
+                DAAL_CHECK_STATUS_VAR(_treeLevelBuildHelper.doNodesSplit(nodeList, nNodes, nodeListNewLevel));
+
+                levelNodeLists.push_back(nodeListNewLevel);
+                levelNodeImpLists.push_back(impListNewLevel);
+
+                DAAL_CHECK_STATUS_VAR(_treeLevelBuildHelper.doLevelPartition(indexedFeatures.getFullData(), nodeList, nNodes, treeOrderLev,
+                                                                             treeOrderLevBuf, nSelectedRows, nFeatures));
+            }
+
+            nNodes = nNodesNewLevel;
+        } // for level
+
+        services::Collection<SharedPtr<algorithmFPType> > binValuesHost(nFeatures);
+        DAAL_CHECK_MALLOC(binValuesHost.data());
+        services::Collection<algorithmFPType *> binValues(nFeatures);
+        DAAL_CHECK_MALLOC(binValues.data());
+
+        for (size_t i = 0; i < nFeatures; i++)
+        {
+            binValuesHost[i] = indexedFeatures.binBorders(i).template get<algorithmFPType>().toHost(ReadWriteMode::readOnly);
+            DAAL_CHECK_MALLOC(binValuesHost[i].get());
+            binValues[i] = binValuesHost[i].get();
+        }
+
+        typename DFTreeConverterType::TreeHelperType mTreeHelper;
+
+        DFTreeConverterType converter;
+        converter.convertToDFDecisionTree(DFTreeRecords, binValues.data(), mTreeHelper, _nClasses);
+
+        mdImpl.add(mTreeHelper._tree, _nClasses);
+
+        DAAL_CHECK_STATUS_VAR(computeResults(mTreeHelper._tree, dataBlock.getBlockPtr(), responseBlock.getBlockPtr(), nSelectedRows, nFeatures,
+                                             oobRows, nOOBRows, oobBufferPerObs, varImpBlock.getBlockPtr(), varImpVariance.get(), iter + 1,
+                                             engines[iter], par));
+
+        DAAL_CHECK_STATUS_VAR(const_cast<NumericTable *>(y)->releaseBlockOfRows(responseBlock));
+    }
+
+    /* Finalize results */
+    if (par.resultsToCompute & (decision_forest::training::computeOutOfBagError | decision_forest::training::computeOutOfBagErrorPerObservation))
+    {
+        BlockDescriptor<algorithmFPType> responseBlock;
+        DAAL_CHECK_STATUS_VAR(const_cast<NumericTable *>(y)->getBlockOfRows(0, nRows, readOnly, responseBlock));
```

Comment:
You should use ReadRows (WriteRows) objects instead of BlockDescriptor to automatically release blocks in case if something fails in between.

Reply:
agree, but these classes have cpu as template param -> need to create some wrapper without cpu dependency for next release